New Examples section with local dev examples (#2374)
astronomer/docs: learn/pycharm-local-dev.md

---
title: "Develop locally with PyCharm"
description: "Integrate the Astro CLI with PyCharm for local development."
id: pycharm-local-dev
sidebar_label: "PyCharm"
---

> This example shows how to set up [PyCharm]((https://www.jetbrains.com/pycharm/)) for local development with Airflow and the [Astro CLI](https://docs.astronomer.io/astro/cli/overview).

Reviewer: ```suggestion
This example shows how to set up [PyCharm](https://www.jetbrains.com/pycharm/) for local development with Airflow and the [Astro CLI](https://docs.astronomer.io/astro/cli/overview).
```

> 

Reviewer: This might be out of scope but do we want to have 1-2 sentences why doing this is beneficial? Or maybe it is self explanatory but something like:

"By setting up your local dev environment you will be able to iterate more quickly when developing Airflow DAGs for example by [debugging interactively with dag.test()](https://docs.astronomer.io/learn/testing-airflow#debug-interactively-with-dagtest)."

## Before you start

> Before using this example, make sure you have:

Reviewer 2: ```suggestion
Before trying this example, make sure you have:
```

- [PyCharm](https://www.jetbrains.com/pycharm/)
- The [Astro CLI](https://docs.astronomer.io/astro/cli/install-cli) installed and a project running locally

## Configure the Python interpreter

> In your PyCharm preferences go to **Build, Execution, Deployment** >> **Docker** and set up the following:

Reviewer: This probably is because I have never used pycharm but I am a little confused what I am supposed to set. 
Is it
"select the correct Docker daemon connection for your set up and add a path mapping for `/Users` to be able to mount files from this folder"

not sure if that setting is correct, I am assuming by mapping Users to Users you are able to create cp statements for files in Users, you are not mounting the whole User directory....? 👀

Author: Your text is correct, although the /Users part is prepopulated, but can be updated. I updated this section to link to the PyCharm docs instead of the screenshot, since they go in more depth and it might vary for different setups.

![Docker settings](/img/examples/pycharm_local_dev_docker_settings.png)

Go to **Project: your-project-name** >> **Python Interpreter** and click the settings gear icon next to Python Interpreter, then click **Add**:

![Interpreter settings](/img/examples/pycharm_local_dev_interpreter.png)

> In the window that appears, click on **Docker** in the left pane, and in the **Image name** field select the Docker image that your Airflow environment is running. Then click **OK**.

Reviewer: ! does this allow you to use the python env from the docker image in the IDE when using dag.test? So you would not need to locally install the provider packages but can still use IDE debugging? 🤔

Author: Hmm...I believe it would? It basically allows you to exec into the containers just like you would from your terminal, but directly from the IDE. So I guess you could probably do that and then run your Python file from there? Not sure whether you can run the file from PyCharm and have it run within Docker?

Reviewer: That looks really useful, I had no idea, thank you! :)

![Image settings](/img/examples/pycharm_local_dev_docker_image.png)

> ## Write Airflow code with PyCharm

Author: This section is second because you need to configure the Python interpreter before you can do this. Will update the descriptions to make that more clear.

Using PyCharm to write Airflow DAGs has multiple advantages like code autocompletion, identifying deprecated or unused imports, and error and warning syntax highlighting.

Here is an example of deprecated imports:

![Deprecated Imports](/img/examples/pycharm_local_dev_deprecated_import.png)

Here is an example of an unused import:

![Unused Imports](/img/examples/pycharm_local_dev_unused_import.png)

> Here is an example error syntax highlighting:

Reviewer 2: ```suggestion
Like with other Python projects, PyCharm will highlight syntax errors in your Airflow code:
```

![Syntax Highlighting](/img/examples/pycharm_local_dev_syntax_highlighting.png)

> Here is an example of code autocompletion and built in definitions:

Reviewer 2: ```suggestion
Lastly, here is an example of PyCharm autocompleting code and showing built-in definitions:
```

![Code Autocompletion](/img/examples/pycharm_local_dev_autocomplete.png)

> ## Manage Docker containers with PyCharm

Author: Hmm....I'm on the fence about this. Generally the flow is you have to set up an interpreter to do anything Python related in PyCharm, then you can write code/do other things. For this example, we use a docker-specific interpreter so that you have the option of interacting with docker containers, but even if you were to remove this, an interpreter is still necessary for the writing code part.

So the flow we have write now is configure an interpreter (and choose one that allows you to do "other things") -> write code -> do "other things" (manage your docker containers). I would lean towards keeping these sections and I'll add more explanation around it, but definitely push back if you think it still doesn't work!

With PyCharm configured to use the Python Interpreter from Docker, you can connect to your Docker containers directly from PyCharm using the ***Services*** pane. If you do not see this pane, try pressing `cmd+8`.

![Services](/img/examples/pycharm_local_dev_docker_services.png)

From the Services pane, start Docker by clicking the green play button and you’ll see the same containers appear as when you run `docker ps`:

![Containers](/img/examples/pycharm_local_dev_containers.png)

You can view logs for the containers by clicking on `/scheduler`, `/triggerer`, `/webserver`, or `/airflow-dev_2cd823-postgres` (this may be a different string based on what the name of the parent directory for your project is):

![Logs](/img/examples/pycharm_local_dev_logs.png)

> You can run Airflow CLI commands by right clicking on `/scheduler` and selecting `Create Terminal` to bash into the container:

Reviewer 2: ```suggestion
Run Airflow CLI commands by right clicking on `/scheduler` and selecting **Create Terminal** to bash into the container:
```
